### DISCUSSION TITLE
Insertion point support?

### DISCUSSION THREAD

**Discussion by dandean** [stephenh/ts-proto#400] [Q&A]
Does `ts-proto` support insertion points?

https://github.com/stephenh/ts-proto/blob/7edc991d83f6351d41b1cc69c40d6ff391dbbd56/protos/google/protobuf/compiler/plugin.proto#L135-L172

If not, do you plan on adding support, or do you have pointers to get started adding this feature on my own?

The use-case I have is that I am building my own service generation. I see that there's already quite a bit of service generation code directly in `ts-proto`, and each is unique and specific in its own way. I don't want to add Yet Another Service Generation option, and it seems like a perfect opportunity for insertion points so that I can write my own plugin which simply amends `ts-proto` output with my own extensions.

To achieve that I think I would just need one insertion point up top for imports, and one insertion point at the bottom for my own generated service additions.

**Comment by dandean** [comment]
It looks like I could probably patch this bit of `main.ts`:

https://github.com/stephenh/ts-proto/blob/main/src/main.ts#L81-L86

And add something like this on line 82 to define an insertion point for imports:

```ts
chunks.push(code`// @@protoc_insertion_point(imports)`);
```

And then add something like this on line 244 to define an insertion point at the end of the file for my own generate services to be inserted:

https://github.com/stephenh/ts-proto/blob/main/src/main.ts#L243-L245

```ts
chunks.push(code`// @@protoc_insertion_point(end_of_file)`);
```

If this would work, one benefit of this would be that all of the existing service-generation options could be pulled out into standalone protoc plugins, which I think would reduce `ts-proto` complexity a bit.

**Reply by dandean** [reply]
Ok cool - I have a hacked up version of `ts-proto` on my machine that's doing this. Here's the output, showing the insertion point, and the code inserted from my own protoc plugin which makes use of the insertion point:

```ts
/* eslint-disable */
import Long from 'long';
import _m0 from 'protobufjs/minimal';

import { MyBaseService } from '@fake/service';
// @@protoc_insertion_point(imports)
```

If you're ok with this feature, I'll work up a PR in the next few days to add these two insertion points.

**Reply by dandean** [reply]
Oh yeah, and here's the output for the end of file insertion point:

```ts
class FooService extends MyBaseService {}
// @@protoc_insertion_point(end_of_file)
```

**Comment by stephenh** [comment]
Hm, interesting, no I hadn't seen the insertion point option/approach before.

FWIW @aikoven recently added the `outputServices=generic-definitions`, which he's using in a framework to do likely similar custom-ish service stuff. (Tangentially, @aikoven wdyt about linking to your project from the ts-proto readme? Maybe either from like a "Community" section or even without the `outputServices` section of the docs as an example for other users to follow your approach to custom services.)

I'd be tempted to look at that first, and hopefully build your custom services on top of that.

If that really doesn't work, I guess I'm not 100% opposed to the insertion point approach, like a `--ts_proto_opt=insertPoints=end_of_file,imports` option...

For my own personal tendencies, it seems a little brittle; although I can see how it makes sense for the `protoc` stuff given their built-in code generators are C++ (!) and so pretty hard to fork/customize/etc. :-)

**Reply by aikoven** [reply]
> wdyt about linking to your project from the ts-proto readme? Maybe either from like a "Community" section or even without the outputServices section of the docs as an example for other users to follow your approach to custom services.

That would be nice! I can make a PR once I have some free time.

Considering the topic of the discussion: AFAIK the core Protobuf already deprecate the insertion point feature, see https://github.com/protocolbuffers/protobuf/pull/6697#issuecomment-837478896

However, I too see it very useful if we could somehow extend the `ts-proto` generator. Maybe through some kind of plugin system, though it sounds really complicated. Another option is to be able to use `ts-proto` as a library, adding custom functionality on top of it.

**Reply by dandean** [reply]
@aikoven Thanks for the link to insertion points being deprecated – I hadn't seen that before.

One thing I like about the path I've been on is that it generates a complete class with all of the methods defined in the protobuf statically on the TypeScript class itself, and because those generated classes are added to the bottom of the file generated by ts-proto there are no imports to worry about.

Following the path you're recommending, I think my plugin will have to instead write to its own files (instead of amending ts-proto's) adjacent to ts-proto files, and resolve all imports associated with the request and response types for each method.

Thanks for the discussion – it's very helpful.

**Comment by stephenh** [comment]
Yeah. Just talking out loud, an earlier version of the ts-poet code generation library used to be more object-based (the current API is basically "give me a bunch of strings, I'll concat them and then auto-organize / de-conflict the imports").

The object-based API was more like "here's a function `Foo`, here's a class `Bar` with methods `zaz`", i.e. the library new more about the structure of the output, which had pros/cons, but it would allow a neat sort of approach like:

```typescript
function myPlugin(output: TsProtoOutput) {
  const file = output.getFile("./simple.ts");
  const foo = file.getClass("FooRpcService");
  const getFoo = foo.getMethod("getFoo");
  foo.setMethod("getFoo", code`
    ...my customization...
    ${getFoo}
    ...my customization...
  `);
  return output;
}
```

And so you could like muck with the files / methods / functions after ts-proto had already made them, but before they go to disk.

Granted, it would take some infrastructure to build, but I think fundamentally wouldn't be too hard/complex.

That said, I do like the simplicity of ideally using the generic-definitions output.
